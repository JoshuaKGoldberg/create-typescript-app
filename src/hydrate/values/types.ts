import { InputValues } from "../../shared/inputs.js";

export interface HydrationInputValues extends InputValues {
	author: string;
	email: string;
}

export interface PartialPackageData {
	author?: { email: string; name: string } | string;
	description?: string;
	email?: string;
	name?: string;
	repository?: string;
}
