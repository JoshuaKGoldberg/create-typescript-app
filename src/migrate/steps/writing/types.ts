export interface Structure {
	[i: string]: Structure | string;
}
