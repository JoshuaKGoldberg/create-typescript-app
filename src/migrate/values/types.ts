import { InputValues } from "../../shared/inputs.js";

export interface MigrationInputValues extends InputValues {
	author: string;
	email: string;
}

export interface PartialPackageData {
	author?: { email: string; name: string } | string;
	description?: string;
	email?: string;
	name?: string;
	repository?: string;
}
