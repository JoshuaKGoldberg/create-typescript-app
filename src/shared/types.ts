export interface AllContributorContributor {
	contributions: string[];
	login: string;
}

export interface AllContributorsData {
	contributors: AllContributorContributor[];
}

export interface PartialPackageData {
	author?: { email: string; name: string } | string;
	dependencies?: Record<string, string>;
	description?: string;
	devDependencies?: Record<string, string>;
	email?: string;
	name?: string;
	repository?: { type: string; url: string } | string;
}

export type InputBase = "everything" | "minimum" | "prompt";

export interface Options {
	author: string | undefined;
	base: InputBase | undefined;
	createRepository: boolean | undefined;
	description: string;
	email: string | undefined;
	excludeCompliance: boolean | undefined;
	excludeContributors: boolean | undefined;
	excludeLintJson: boolean | undefined;
	excludeLintKnip: boolean | undefined;
	excludeLintMd: boolean | undefined;
	excludeLintPackageJson: boolean | undefined;
	excludeLintPackages: boolean | undefined;
	excludeLintPerfectionist: boolean | undefined;
	excludeLintSpelling: boolean | undefined;
	excludeLintYml: boolean | undefined;
	excludeReleases: boolean | undefined;
	excludeRenovate: boolean | undefined;
	excludeTests: boolean | undefined;
	funding: string | undefined;
	owner: string;
	repository: string;
	skipGitHubApi: boolean;
	skipInstall: boolean | undefined;
	skipRemoval: boolean | undefined;
	skipRestore: boolean | undefined;
	skipUninstall: boolean | undefined;
	title: string;
}
