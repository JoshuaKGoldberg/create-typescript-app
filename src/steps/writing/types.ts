export interface Structure {
	[i: string]: string | Structure;
}
