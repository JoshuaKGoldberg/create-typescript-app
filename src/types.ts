export interface GreetOptions {
	logger?: (message: string) => void;
	message: string;
	times?: number;
}
