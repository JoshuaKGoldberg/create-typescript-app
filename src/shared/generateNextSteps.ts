import { OutroGroup } from "./cli/outro.js";
import { Options } from "./types.js";

export function generateNextSteps(options: Options): OutroGroup[] {
	const lines = [
		joinedMessage(
			"enable the",
			[
				options.excludeTests || ["Codecov", "https://github.com/apps/codecov"],
				options.excludeRenovate || [
					"Renovate",
					"https://github.com/apps/renovate",
				],
			],
			"GitHub app",
		),
		joinedMessage(
			"populate the",
			[
				(options.excludeAllContributors && options.excludeReleases) || [
					"ACCESS_TOKEN",
					"a GitHub PAT with repo and workflow permissions",
				],
				options.excludeReleases || [
					"NPM_TOKEN",
					"an npm access token with automation permissions",
				],
			],
			"secret",
		),
	].filter((line): line is string => !!line);

	switch (lines.length) {
		case 0:
			return [];
		case 1:
			return [{ label: `Be sure to ${lines[0]}` }];
		default:
			return [
				{
					label: "Be sure to:",
					lines: lines.map((line) => `- ${line}`),
				},
			];
	}
}

const itemBreak = "\n   - ";

function joinedMessage(
	prefix: string,
	entries: ([string, string] | true)[],
	suffix: string,
) {
	const realEntries = entries.filter(
		(entry): entry is [string, string] => entry !== true,
	);

	switch (realEntries.length) {
		case 0:
			return undefined;
		case 1:
			return `${prefix} ${realEntries[0][0]} ${suffix} (${realEntries[0][1]}).`;
		default:
			return `${prefix} ${suffix}s: ${itemBreak}${realEntries
				.map((entry) => `${entry[0]} (${entry[1]})`)
				.join(itemBreak)}`;
	}
}
