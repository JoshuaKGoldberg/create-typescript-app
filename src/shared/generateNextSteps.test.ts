import { describe, expect, test } from "vitest";

import { generateNextSteps } from "./generateNextSteps.js";
import { Options } from "./types.js";

const options = {
	access: "public",
	base: "everything",
	description: "Test description.",
	directory: ".",
	email: {
		github: "github@email.com",
		npm: "npm@email.com",
	},
	logo: undefined,
	mode: "create",
	owner: "TestOwner",
	repository: "test-repository",
	title: "Test Title",
} satisfies Options;

describe("generateNextSteps", () => {
	for (const excludeAllContributors of [false, true]) {
		for (const excludeReleases of [false, true]) {
			for (const excludeRenovate of [false, true]) {
				for (const excludeTests of [false, true]) {
					test(
						// eslint-disable-next-line vitest/valid-title
						JSON.stringify({
							excludeAllContributors,
							excludeReleases,
							excludeRenovate,
							excludeTests,
						}),
						() => {
							expect(
								generateNextSteps({
									...options,
									excludeAllContributors,
									excludeReleases,
									excludeRenovate,
									excludeTests,
								}),
							).toMatchSnapshot();
						},
					);
				}
			}
		}
	}
});
