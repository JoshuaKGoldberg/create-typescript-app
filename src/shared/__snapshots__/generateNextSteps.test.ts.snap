// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateNextSteps > {"excludeAllContributors":false,"excludeReleases":false,"excludeRenovate":false,"excludeTests":false} 1`] = `
[
  {
    "label": "Be sure to:",
    "lines": [
      "- enable the GitHub apps: 
   - Codecov (https://github.com/apps/codecov)
   - Renovate (https://github.com/apps/renovate)",
      "- populate the secrets: 
   - ACCESS_TOKEN (a GitHub PAT with repo and workflow permissions)
   - NPM_TOKEN (an npm access token with automation permissions)",
    ],
  },
]
`;

exports[`generateNextSteps > {"excludeAllContributors":false,"excludeReleases":false,"excludeRenovate":false,"excludeTests":true} 1`] = `
[
  {
    "label": "Be sure to:",
    "lines": [
      "- enable the Renovate GitHub app (https://github.com/apps/renovate).",
      "- populate the secrets: 
   - ACCESS_TOKEN (a GitHub PAT with repo and workflow permissions)
   - NPM_TOKEN (an npm access token with automation permissions)",
    ],
  },
]
`;

exports[`generateNextSteps > {"excludeAllContributors":false,"excludeReleases":false,"excludeRenovate":true,"excludeTests":false} 1`] = `
[
  {
    "label": "Be sure to:",
    "lines": [
      "- enable the Codecov GitHub app (https://github.com/apps/codecov).",
      "- populate the secrets: 
   - ACCESS_TOKEN (a GitHub PAT with repo and workflow permissions)
   - NPM_TOKEN (an npm access token with automation permissions)",
    ],
  },
]
`;

exports[`generateNextSteps > {"excludeAllContributors":false,"excludeReleases":false,"excludeRenovate":true,"excludeTests":true} 1`] = `
[
  {
    "label": "Be sure to populate the secrets: 
   - ACCESS_TOKEN (a GitHub PAT with repo and workflow permissions)
   - NPM_TOKEN (an npm access token with automation permissions)",
  },
]
`;

exports[`generateNextSteps > {"excludeAllContributors":false,"excludeReleases":true,"excludeRenovate":false,"excludeTests":false} 1`] = `
[
  {
    "label": "Be sure to:",
    "lines": [
      "- enable the GitHub apps: 
   - Codecov (https://github.com/apps/codecov)
   - Renovate (https://github.com/apps/renovate)",
      "- populate the ACCESS_TOKEN secret (a GitHub PAT with repo and workflow permissions).",
    ],
  },
]
`;

exports[`generateNextSteps > {"excludeAllContributors":false,"excludeReleases":true,"excludeRenovate":false,"excludeTests":true} 1`] = `
[
  {
    "label": "Be sure to:",
    "lines": [
      "- enable the Renovate GitHub app (https://github.com/apps/renovate).",
      "- populate the ACCESS_TOKEN secret (a GitHub PAT with repo and workflow permissions).",
    ],
  },
]
`;

exports[`generateNextSteps > {"excludeAllContributors":false,"excludeReleases":true,"excludeRenovate":true,"excludeTests":false} 1`] = `
[
  {
    "label": "Be sure to:",
    "lines": [
      "- enable the Codecov GitHub app (https://github.com/apps/codecov).",
      "- populate the ACCESS_TOKEN secret (a GitHub PAT with repo and workflow permissions).",
    ],
  },
]
`;

exports[`generateNextSteps > {"excludeAllContributors":false,"excludeReleases":true,"excludeRenovate":true,"excludeTests":true} 1`] = `
[
  {
    "label": "Be sure to populate the ACCESS_TOKEN secret (a GitHub PAT with repo and workflow permissions).",
  },
]
`;

exports[`generateNextSteps > {"excludeAllContributors":true,"excludeReleases":false,"excludeRenovate":false,"excludeTests":false} 1`] = `
[
  {
    "label": "Be sure to:",
    "lines": [
      "- enable the GitHub apps: 
   - Codecov (https://github.com/apps/codecov)
   - Renovate (https://github.com/apps/renovate)",
      "- populate the secrets: 
   - ACCESS_TOKEN (a GitHub PAT with repo and workflow permissions)
   - NPM_TOKEN (an npm access token with automation permissions)",
    ],
  },
]
`;

exports[`generateNextSteps > {"excludeAllContributors":true,"excludeReleases":false,"excludeRenovate":false,"excludeTests":true} 1`] = `
[
  {
    "label": "Be sure to:",
    "lines": [
      "- enable the Renovate GitHub app (https://github.com/apps/renovate).",
      "- populate the secrets: 
   - ACCESS_TOKEN (a GitHub PAT with repo and workflow permissions)
   - NPM_TOKEN (an npm access token with automation permissions)",
    ],
  },
]
`;

exports[`generateNextSteps > {"excludeAllContributors":true,"excludeReleases":false,"excludeRenovate":true,"excludeTests":false} 1`] = `
[
  {
    "label": "Be sure to:",
    "lines": [
      "- enable the Codecov GitHub app (https://github.com/apps/codecov).",
      "- populate the secrets: 
   - ACCESS_TOKEN (a GitHub PAT with repo and workflow permissions)
   - NPM_TOKEN (an npm access token with automation permissions)",
    ],
  },
]
`;

exports[`generateNextSteps > {"excludeAllContributors":true,"excludeReleases":false,"excludeRenovate":true,"excludeTests":true} 1`] = `
[
  {
    "label": "Be sure to populate the secrets: 
   - ACCESS_TOKEN (a GitHub PAT with repo and workflow permissions)
   - NPM_TOKEN (an npm access token with automation permissions)",
  },
]
`;

exports[`generateNextSteps > {"excludeAllContributors":true,"excludeReleases":true,"excludeRenovate":false,"excludeTests":false} 1`] = `
[
  {
    "label": "Be sure to enable the GitHub apps: 
   - Codecov (https://github.com/apps/codecov)
   - Renovate (https://github.com/apps/renovate)",
  },
]
`;

exports[`generateNextSteps > {"excludeAllContributors":true,"excludeReleases":true,"excludeRenovate":false,"excludeTests":true} 1`] = `
[
  {
    "label": "Be sure to enable the Renovate GitHub app (https://github.com/apps/renovate).",
  },
]
`;

exports[`generateNextSteps > {"excludeAllContributors":true,"excludeReleases":true,"excludeRenovate":true,"excludeTests":false} 1`] = `
[
  {
    "label": "Be sure to enable the Codecov GitHub app (https://github.com/apps/codecov).",
  },
]
`;

exports[`generateNextSteps > {"excludeAllContributors":true,"excludeReleases":true,"excludeRenovate":true,"excludeTests":true} 1`] = `[]`;
